File: typescript/tests/shared/test-utils.ts
import { expect } from "vitest";
import { ApiClient } from "@/api/client";
import type { Context } from "@/tools/types";
import { MemoryCache } from "@/lib/utils/cache/MemoryCache";

export const API_BASE_URL = process.env.TEST_API_BASE_URL || "http://localhost:8010";
export const API_TOKEN = process.env.TEST_API_TOKEN;
export const TEST_ORG_ID = process.env.TEST_ORG_ID;
export const TEST_PROJECT_ID = process.env.TEST_PROJECT_ID;

export interface CreatedResources {
	featureFlags: number[];
	insights: number[];
	dashboards: number[];
}

export function validateEnvironmentVariables() {
	if (!API_TOKEN) {
		throw new Error("TEST_API_TOKEN environment variable is required");
	}

	if (!TEST_ORG_ID) {
		throw new Error("TEST_ORG_ID environment variable is required");
	}

	if (!TEST_PROJECT_ID) {
		throw new Error("TEST_PROJECT_ID environment variable is required");
	}
}

export function createTestClient(): ApiClient {
	return new ApiClient({
		apiToken: API_TOKEN!,
		baseUrl: API_BASE_URL,
	});
}

export function createTestContext(client: ApiClient): Context {
	const cache = new MemoryCache<any>("test-user");

	const context: Context = {
		api: client,
		cache,
		env: {} as any,
		getProjectId: async () => {
			const projectId = await cache.get("projectId");
			if (!projectId) {
				throw new Error("No active project set");
			}
			return projectId;
		},
		getOrgID: async () => {
			const orgId = await cache.get("orgId");
			if (!orgId) {
				throw new Error("No active organization set");
			}
			return orgId;
		},
		getDistinctId: async () => {
			const distinctId = await cache.get("distinctId");
			return distinctId || "";
		},
	};

	return context;
}

export async function setActiveProjectAndOrg(context: Context, projectId: string, orgId: string) {
	const cache = context.cache;
	await cache.set("projectId", projectId);
	await cache.set("orgId", orgId);
}

export async function cleanupResources(
	client: ApiClient,
	projectId: string,
	resources: CreatedResources,
) {
	for (const flagId of resources.featureFlags) {
		try {
			await client.featureFlags({ projectId }).delete({ flagId });
		} catch (error) {
			console.warn(`Failed to cleanup feature flag ${flagId}:`, error);
		}
	}
	resources.featureFlags = [];

	for (const insightId of resources.insights) {
		try {
			await client.insights({ projectId }).delete({ insightId });
		} catch (error) {
			console.warn(`Failed to cleanup insight ${insightId}:`, error);
		}
	}
	resources.insights = [];

	for (const dashboardId of resources.dashboards) {
		try {
			await client.dashboards({ projectId }).delete({ dashboardId });
		} catch (error) {
			console.warn(`Failed to cleanup dashboard ${dashboardId}:`, error);
		}
	}
	resources.dashboards = [];
}

export function parseToolResponse(result: any) {
	expect(result.content).toBeDefined();
	expect(result.content[0].type).toBe("text");
	return JSON.parse(result.content[0].text);
}

export function generateUniqueKey(prefix: string): string {
	return `${prefix}-${Date.now()}-${Math.random().toString(36).substring(7)}`;
}

export const SAMPLE_HOGQL_QUERIES = {
	pageviews: {
		kind: "DataVisualizationNode" as const,
		source: {
			kind: "HogQLQuery" as const,
			query: "SELECT event, count() AS event_count FROM events WHERE timestamp >= now() - INTERVAL 7 DAY AND event = '$pageview' GROUP BY event ORDER BY event_count DESC LIMIT 10",
			filters: {
				dateRange: {
					date_from: "-7d",
					date_to: "-1d",
				},
			},
		},
	},
	topEvents: {
		kind: "DataVisualizationNode" as const,
		source: {
			kind: "HogQLQuery" as const,
			query: "SELECT event, count() AS event_count FROM events WHERE timestamp >= now() - INTERVAL 7 DAY GROUP BY event ORDER BY event_count DESC LIMIT 10",
			filters: {
				dateRange: {
					date_from: "-7d",
					date_to: "-1d",
				},
			},
		},
	},
};

File: python/tests/shared/test_utils.py
import json
import os
import random
import string
import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from api.client import ApiClient, ApiConfig
from lib.config import PostHogToolConfig
from lib.utils.cache.memory_cache import MemoryCache
from schema.query import DateRange, HogQLFilters, HogQLQuery, InsightQuery
from tools.types import Context, ToolResult


def load_env_test_file():
    """Load environment variables from .env.test file."""
    # Look for .env.test in the project root (parent of python directory)
    env_test_path = Path(__file__).parent.parent.parent.parent / ".env.test"
    if env_test_path.exists():
        with open(env_test_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#"):
                    key, value = line.split("=", 1)
                    # Remove quotes if present
                    value = value.strip("\"'")
                    os.environ[key] = value


# Load environment variables from .env.test
load_env_test_file()

API_BASE_URL = os.getenv("TEST_POSTHOG_API_BASE_URL", "http://localhost:8010")
PERSONAL_API_KEY = os.getenv("TEST_POSTHOG_PERSONAL_API_KEY", "")
TEST_ORG_ID = os.getenv("TEST_ORG_ID", "")
TEST_PROJECT_ID = os.getenv("TEST_PROJECT_ID", "")


@dataclass
class CreatedResources:
    feature_flags: list[int]
    insights: list[int]
    dashboards: list[int]

    def __init__(self):
        self.feature_flags = []
        self.insights = []
        self.dashboards = []


def validate_environment_variables():
    """Validate that required environment variables are set."""
    if not PERSONAL_API_KEY:
        raise ValueError("TEST_POSTHOG_PERSONAL_API_KEY environment variable is required")

    if not TEST_ORG_ID:
        raise ValueError("TEST_ORG_ID environment variable is required")

    if not TEST_PROJECT_ID:
        raise ValueError("TEST_PROJECT_ID environment variable is required")


def create_test_client() -> ApiClient:
    """Create a test API client."""
    return ApiClient(ApiConfig(personal_api_key=PERSONAL_API_KEY, base_url=API_BASE_URL))


def create_test_context(client: ApiClient) -> Context:
    """Create a test context with mocked cache and helper functions."""
    cache = MemoryCache()
    config = PostHogToolConfig(
        personal_api_key=PERSONAL_API_KEY,
        api_base_url=API_BASE_URL,
        inkeep_api_key=os.getenv("INKEEP_API_KEY"),
        dev=os.getenv("DEV", "false").lower() in ("true", "1", "yes"),
    )

    async def get_project_id() -> str:
        project_id = await cache.get("project_id")
        if project_id is None:
            raise Exception("No active project set")
        return project_id

    async def get_org_id() -> str:
        org_id = await cache.get("org_id")
        if org_id is None:
            raise Exception("No active organization set")
        return org_id

    async def get_distinct_id() -> str:
        distinct_id = await cache.get("distinct_id")
        return distinct_id or uuid.uuid4().hex

    return Context(
        api=client,
        cache=cache,
        config=config,
        get_project_id=get_project_id,
        get_org_id=get_org_id,
        get_distinct_id=get_distinct_id,
    )


async def set_active_project_and_org(context: Context, project_id: str, org_id: str):
    """Set active project and organization in the cache."""
    await context.cache.set("project_id", project_id)
    await context.cache.set("org_id", org_id)


async def cleanup_resources(client: ApiClient, project_id: str, resources: CreatedResources):
    """Clean up test resources."""

    # Clean up feature flags
    for flag_id in resources.feature_flags:
        try:
            await client.feature_flags(project_id).delete(flag_id)
        except Exception as e:
            print(f"Failed to cleanup feature flag {flag_id}: {e}")
    resources.feature_flags.clear()

    # Clean up insights
    for insight_id in resources.insights:
        try:
            await client.insights(project_id).delete(insight_id)
        except Exception as e:
            print(f"Failed to cleanup insight {insight_id}: {e}")
    resources.insights.clear()

    # Clean up dashboards
    for dashboard_id in resources.dashboards:
        try:
            await client.dashboards(project_id).delete(dashboard_id)
        except Exception as e:
            print(f"Failed to cleanup dashboard {dashboard_id}: {e}")
    resources.dashboards.clear()


def parse_tool_result(result: ToolResult) -> Any:
    """Parse the JSON response from a tool execution."""
    return json.loads(result.content)


def generate_unique_key(prefix: str) -> str:
    """Generate a unique key for testing."""
    timestamp = int(time.time() * 1000)
    random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=7))
    return f"{prefix}-{timestamp}-{random_suffix}"


def create_sample_queries() -> dict[str, InsightQuery]:
    pageviews_query = InsightQuery(
        kind="DataVisualizationNode",
        source=HogQLQuery(
            kind="HogQLQuery",
            query="SELECT event, count() AS event_count FROM events WHERE timestamp >= now() - INTERVAL 7 DAY AND event = '$pageview' GROUP BY event ORDER BY event_count DESC LIMIT 10",
            filters=HogQLFilters(dateRange=DateRange(date_from="-7d", date_to="-1d")),
        ),
    )

    top_events_query = InsightQuery(
        kind="DataVisualizationNode",
        source=HogQLQuery(
            kind="HogQLQuery",
            query="SELECT event, count() AS event_count FROM events WHERE timestamp >= now() - INTERVAL 7 DAY GROUP BY event ORDER BY event_count DESC LIMIT 10",
            filters=HogQLFilters(dateRange=DateRange(date_from="-7d", date_to="-1d")),
        ),
    )

    return {"pageviews": pageviews_query, "top_events": top_events_query}


SAMPLE_HOGQL_QUERIES = create_sample_queries()


# Sample feature flag filters for testing
SAMPLE_FEATURE_FLAG_FILTERS = {"groups": [{"properties": [], "rollout_percentage": 100}]}


SAMPLE_FEATURE_FLAG_FILTERS_WITH_PROPERTIES = {
    "groups": [
        {
            "properties": [{"key": "email", "value": "test@posthog.com", "operator": "exact", "type": "person"}],
            "rollout_percentage": 50,
        }
    ]
}
